pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,776 @@
+
+                    self._prep.append(obj)
+                else:
+                    self._ops.append(obj)
+
+            elif isinstance(obj, MeasurementProcess):
+                # measurement process
+
+                if obj.return_type is qml.operation.Probability:
+                    self._obs.append((obj, obj))
+                    self._output_dim += 2 ** len(obj.wires)
+
+                elif "owns" in info:
+                    # TODO: remove the following line once devices
+                    # have been refactored to no longer use obs.return_type.
+                    # Monkeypatch the observable to have the same return
+                    # type as the measurement process.
+                    info["owns"].return_type = obj.return_type
+                    info["owns"].do_check_domain = False
+
+                    self._obs.append((obj, info["owns"]))
+                    self._output_dim += 1
+
+                    if obj.return_type is qml.operation.Sample:
+                        self.is_sampled = True
+
+            elif isinstance(obj, qml.operation.Observable) and "owner" not in info:
+                raise ValueError(f"Observable {obj} does not have a measurement type specified.")
+
+        self._update()
+
+    def _update_circuit_info(self):
+        """Update circuit metadata"""
+        self.wires = qml.wires.Wires.all_wires(
+            [op.wires for op in self.operations + self.observables]
+        )
+        self.num_wires = len(self.wires)
+
+    def _update_par_info(self):
+        """Update the parameter information dictionary"""
+        param_count = 0
+
+        for obj in self.operations + self.observables:
+
+            for p in range(len(obj.data)):
+                info = self._par_info.get(param_count, {})
+                info.update({"op": obj, "p_idx": p})
+
+                self._par_info[param_count] = info
+                param_count += 1
+
+    def _update_trainable_params(self):
+        """Set the trainable parameters"""
+        self._trainable_params = set(self._par_info)
+
+    def _update(self):
+        """Update all internal tape metadata regarding processed operations and observables"""
+        self._graph = None
+        self._update_circuit_info()
+        self._update_par_info()
+        self._update_trainable_params()
+
+    def _expand(self, stop_at=None):
+        """Expand all operations and tapes in the processed queue.
+
+        Args:
+            stop_at (Sequence[str]): Sequence of PennyLane operation or tape names.
+                An operation or tape appearing in this list of names is not expanded.
+        """
+        new_tape = self.__class__()
+        stop_at = stop_at or []
+
+        for op in self.operations:
+            if op.name in stop_at:
+                new_tape._ops.append(op)
+                continue
+
+            t = op if isinstance(op, QuantumTape) else op.expand()
+            new_tape._prep += t._prep
+            new_tape._ops += t._ops
```

Comment:
I don't think so. Currently, an `Operation` is either an operation (`RX`, `QubitUnitary`) _or_ a state prep (`QubitStateVector`, `BasisState`). I'm not sure it is possible for an operation to be both, since the distinction is made regarding the object as a whole.

The _only_ place where it becomes a bit tricky is decompositions of state preparations. For example, `BasisState` decomposes to a sequence of `PauliX` operations, which using the current definition, is no longer a state prep.

I think this motivates additional bookkeeping for marking state preps (probably better done in a separate PR). Maybe all operations have a boolean attribute that can be set as True/False, indicating whether they are part of a state prep or not?

Reply:
the presence of an op in the state prep subtape should probably be the main indicator?

Reply:
By construction, I don't think could ever happen.

For example:

```python
class BasisState(operation):
    def expand(self):
        with QuantumTape() as tape:
            [qml.PauliX(w) for w, s in zip(self.wires, self.data[0]) if s == 1]
        return tape
```

The tape returned by `BasisState.expand()` will have PauliX gates in the `tape._ops` list, and nothing in the `tape._prep` list. I think we will have to somehow 'mark' operations as state preps, so that the tape knows where to put them.

Reply:
:thinking: